### DISCUSSION TITLE
Multiple Tabs for same user

### DISCUSSION THREAD

**Discussion by paulghobeira** [keycloak/keycloak#23092] [Q&A]
Hello,

I have the below case and I'm wondering if there is a solution for it:

1. User logs in first tab.
2. User opens a page in another tab and leaves it to work on the first tab.
3. The first tab is always authenticated and the access token can be used.
4. Once the user tries to use the second tab, both the access token and refresh token are expired. (in this case I logout the user and redirect them to the login page)

How can I make sure that the user stays authenticated if at least one of the tabs is being used?

**Comment by justin-tay** [comment]
There are two approaches, either using a backend with an application session cookie, or using a service worker. Either of these approaches will be shared across multiple tabs.

You can have a separate backend to be the confidential OAuth client and manage the tokens from Keycloak like https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#name-backend-for-frontend-bff-pr where the application session is actually managed using a http only, secure session cookie and hence the issue with multiple tabs won't come up, since cookies are shared. Of course appropriate CSRF mitigation measures need to be taken with this approach due to the use of a cookie to store the application session.

If having a backend isn't an option probably using a service worker approach https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#name-acquiring-tokens-from-a-ser can be considered.
